### ISSUE TITLE
Crashe in Qemu on Mac when hardware virtualization is enabled

### ISSUE THREAD

**Issue by VincentBott** [managarm/managarm#247]
Hello,

Managarm crashes, when trying to boot it, using a nightly build with Qemu, in MacOsX10.14

I get the following error: 

```

thor: Discovering PCI devices
    Bus: 0, slot 0

            Vendor/device: 8086.1237.2, class: 6.0.0
    Bus: 0, slot 1

            Vendor/device: 8086.7000.0, class: 6.1.0

            Vendor/device: 8086.7010.0, class: 1.1.80
            I/O space BAR #4 at 0xc0e0, length: 16 ports

            Vendor/device: 8086.7113.3, class: 6.80.0
            Interrupt: INTA (routed to io-apic.0:9)
    Bus: 0, slot 2

            Vendor/device: 1af4.1050.1, class: 3.0.0
            MSI-X capability
            Vendor-specific capability
            Vendor-specific capability
            Vendor-specific capability
            Vendor-specific capability
            Vendor-specific capability
            32-bit memory BAR #0 at 0xfe000000, length: 8388608 bytes
qemu-system-x86_64: Error: HV_ERROR
Abort trap: 6
```

I use this command to launch Qemu: 

`
qemu-system-x86_64 -accel hvf  -m 1024 -device piix3-usb-uhci -device usb-kbd -device usb-tablet -drive id=hdd,file=/Users/Vincent/Downloads/image,format=raw,if=none -device virtio-blk-pci,drive=hdd -vga virtio -debugcon stdio -cpu host
`

Many thanks in advance.

**Comment by qookei** [comment]
Hmm, it would appear we somehow triggered a hypervisor error in the PCI code. Could you check whether this happens on a [more recent build](https://ci.managarm.org/repos/files/managarm/2021-02-07T16:22:14/image.xz)? (the image linked in the readme is quite outdated)

**Comment by VincentBott** [comment]
> Hmm, it would appear we somehow triggered a hypervisor error in the PCI code. Could you check whether this happens on a [more recent build](https://ci.managarm.org/repos/files/managarm/2021-02-07T16:22:14/image.xz)? (the image linked in the readme is quite outdated)

I tried the image you linked to, and it still crashes.  With the following error(slightly different):


`thor: Discovering PCI devices
    Segment: 0, bus: 0, slot 0
        Function 0: Device (Decodes IO) (Decodes Memory)
            Vendor/device: 8086.1237.2, class: 6.0.0
    Segment: 0, bus: 0, slot 1
        Function 0: Device (Decodes IO) (Decodes Memory)
            Vendor/device: 8086.7000.0, class: 6.1.0
        Function 1: Device (Decodes IO) (Decodes Memory)
            Vendor/device: 8086.7010.0, class: 1.1.80
            I/O space BAR #4 at 0xc0e0, length: 16 ports
        Function 3: Device (Decodes IO) (Decodes Memory)
            Vendor/device: 8086.7113.3, class: 6.80.0
            Interrupt: INTA (routed to io-apic.0:9)
    Segment: 0, bus: 0, slot 2
        Function 0: Device (Decodes IO) (Decodes Memory)
            Vendor/device: 1af4.1050.1, class: 3.0.0
            MSI-X capability
            Vendor-specific capability
            Vendor-specific capability
            Vendor-specific capability
            Vendor-specific capability
            Vendor-specific capability
            32-bit memory BAR #0 at 0xfe000000, length: 8388608 bytes
qemu-system-x86_64: Error: HV_ERROR
Abort trap: 6`

**Comment by ArsenArsen** [comment]
This looks to be a macOS + qemu issue. `HV_ERROR` comes from Apples hypervisor framework.

**Comment by VincentBott** [comment]
> This looks to be a macOS + qemu issue. `HV_ERROR` comes from Apples hypervisor framework.

I converted the Managarm-image to a VirtualBox-file, but it still crashes, when trying to boot.  Now with a different error: 
<img width="445" alt="Managarm VirtualBox" src="https://user-images.githubusercontent.com/41168811/107640481-a4176400-6c72-11eb-9217-7ed40748134a.png">

**Comment by qookei** [comment]
Looking at this again, I think this is because we don't disable memory decode before trying to determine the BAR size (so we end up writing an illegal address into the BAR, which confuses the hypervisor).

**Comment by no92** [comment]
Probably fixed by #542.
